File: Test_cot_se.py
import math
from math import fabs
from math import pi
import random

def cot_se_p(angle):
    x = (angle/180)*pi;
    x1 = float(x)
    x2 = float(x)
    g = 0
    t = x1
    n = 1
    while (fabs(t) >= 1e-10):
        g += t
        n += 1
        t = -t * x1 * x1 / (2 * n - 1) / (2 * n - 2)   #sin

    cosTotal  = 1
    count = 2
    term = 1
    #x=float(x)
    while abs(term) > 1e-20:
        term *= (-x2 * x2)/( count * (count-1) )
        cosTotal += term
        count += 2                                   #cos
        #print("%1d  %22.17e" % (count, term))
    g = cosTotal/g
    return g


def test_cot():
    num = 1000
    test_number = [0] * num
    sum = 0
    flag = 0  # 判断误差是否大于0.001的标志，默认为0，小于0.001
    for i in range(num):  # 随机选择1000个随机数
        test_number[i] = random.uniform(0, 360)
    for i in range(num):
        sta_val = 1 / math.tan(test_number[i] / 180 * pi)  # 标准值
        tes_val = cot_se_p(test_number[i])  # 需要测试的值
        error = abs(sta_val - tes_val)  # 误差
        if error > 0.001:
            flag = 1
            tes_val = 1 / math.tan(test_number[i] / 180 * pi)
            error = abs(sta_val - tes_val)
            if error <= 0.001:
                flag = 0
        sum = sum + error
    avgError = sum / num  # 平均误差
    return flag, avgError


if __name__ == '__main__':
    flag, avgError = test_cot()
    print('flag = ', flag)
    print('avgError = ', avgError)

File: Test_cot_se.m
function [avgError,flag]=Test_cot_se()
a=round(rand(1,1000)*360);%随机产生一个1*1000(double)数列，其中的值范围：0-360
%b=tand(a);%以角度制计算tan函数，tand是MATLAB固有函数
%t=0;
T=length(a);
%------------------------------------------------
%进行1000次的计算误差分析

  %将标准库中的1/tand()与编写的cot_se()运算值进行误差计算：
     %对每一次误差计算的操作：
          %(1)两者误差值大于设定标准误差，即cot_se()计算错误，flag++
          %(2)两者误差值大于设定标准误差，即cot_se()计算无误
     %如上的操作重复1000（T）次，最终误差计算=测试计算中发生错误的次数/总测试数

for i=1:T %比较，计算错误的个数
    b(i)=1/tand(a(i));%标准值
    c(i) = cot_se( a(i) );%测试值
    if abs(b(i))==Inf&&abs(c(i))==Inf %考虑a(i)=90的情况
         erro(i)=0;
     else
      erro(i)=abs(b(i)-c(i));%计算每次误差=|标准值-测试值|
     end
    erro(i)=abs(b(i)-c(i));
end
for j=1:T
if erro(j)>0.001
      flag=1;
      break;
  else
      flag=0;
end
end
 avgError=sum(erro)/T;
end
